### PR TITLE
EVG-13893 Add Known Issue display status to task query

### PR DESCRIPTION
**Pull request by malikchaya2**
[EVG-13893](https://jira.mongodb.org/browse/EVG-13893)

This adds "known-issue" as the display status if there is a task annotation in the task annotation collection that matches the taskId, execution, and has at least one issue in the issue list. 

**Testing**
Tested with both a unit test, and by pushing the changes to staging and querying PatchTasks for  patchId: "6049134597b1d353b0ea9ce5" which has a task with a status of "known-issue". 

image

[Spruce changes](https://github.com/evergreen-ci/spruce/pull/753)

### CODE REVIEW

File: rest/model/task.go [hunk 1]
```diff
@@ -275,6 +275,14 @@ func (at *APITask) BuildFromService(t interface{}) error {
 			}
 		}
 
+		isKnown, err := v.IsKnownFailure()
+		if err != nil {
+			return errors.Wrapf(err, "Error checking if the task is a known issue: %v", err.Error())
```
Comment (ablack12) [inline rest/model/task.go:280]: returning the error twice in this statement

File: rest/model/task.go [hunk 2]
```diff
@@ -327,6 +335,17 @@ func (at *APITask) BuildFromService(t interface{}) error {
 
 // ToService returns a service layer task using the data from the APITask.
 func (ad *APITask) ToService() (interface{}, error) {
+	if *ad.Status == evergreen.TaskKnownIssue {
```
Comment (ablack12) [inline rest/model/task.go:338]: rather than accessing string pointers directly, we typically use utility.FromStringPtr for safety

File: rest/model/task.go [hunk 3]
```diff
@@ -327,6 +335,17 @@ func (at *APITask) BuildFromService(t interface{}) error {
...
+		task, err := task.FindOneId(*ad.Id)
+		if err != nil {
+			return "", errors.Wrapf(err, "problem finding task %s", *ad.Id)
+		}
+		if task == nil {
+			return "", errors.Errorf("no task found %s", *ad.Id)
```
Comment (ablack12) [inline rest/model/task.go:344]: Can this error message be worded differently? "task '%s' not found"  reads more like a sentence

**Review by ablack12 — CHANGES_REQUESTED** [review]

File: rest/model/task.go [hunk 4]
```diff
@@ -275,6 +275,14 @@ func (at *APITask) BuildFromService(t interface{}) error {
 			}
 		}
 
+		isKnown, err := v.IsKnownFailure()
+		if err != nil {
+			return errors.Wrapf(err, "Error checking if the task is a known issue: %v", err.Error())
+		}
+		if isKnown {
+			v.Status = evergreen.TaskKnownIssue
```
Comment (ablack12) [inline rest/model/task.go:283]: I'm concerned about making this change here, because BuildFromService is used for API routes as well. I think either this should be done in the resolver, or there should be an additional APITask field called IsKnownIssue or something (then this information could still return whether it's a known issue to API routes without changing what users might currently expect)

File: model/task/task.go [hunk 5]
```diff
@@ -2741,6 +2742,37 @@ func GetTasksByVersion(versionID string, sortBy []TasksSortOrder, statuses []str
 					tempParentKey: []interface{}{},
 				},
 			},
+			// get any annotation that has at least one issue
```
Comment (ablack12) [inline model/task/task.go:2745]: Do we need annotation information everywhere we use this query? If we don't we can pass in a boolean and only add this lookup to the pipeline if we need to.

File: model/task/task.go [hunk 6]
```diff
@@ -2741,6 +2742,37 @@ func GetTasksByVersion(versionID string, sortBy []TasksSortOrder, statuses []str
...
+										{
+											"$eq": []string{"$execution", "$$task_annotation_execution"},
+										},
+										{"$or": []bson.M{
+											{
+												annotations.IssuesKey: bson.M{"$gt": []interface{}{"$" + annotations.IssuesKey, 0}},
+											},
+
+											{
+												annotations.IssuesKey: bson.M{"$ne": []interface{}{"$" + annotations.IssuesKey, bson.TypeNull}},
```
Comment (ablack12) [inline model/task/task.go:2767]: Doesn't this or case mean that if Issues is an empty list we'll still return it?

File: rest/data/task_test.go [hunk 7]
```diff
@@ -81,6 +82,25 @@ func (s *TaskConnectorFetchByIdSuite) TestFindByIdAndExecution() {
 	}
 }
 
+func (s *TaskConnectorFetchByIdSuite) TestFindByVersion() {
```
Comment (ablack12) [inline rest/data/task_test.go:85]: Can this test more cases? (A case where annotation exists but Issues is the empty list, Issues is nil,  no annotation exists, etc)

**Review by ablack12 — CHANGES_REQUESTED** [review]

File: model/task/task.go (hunk 5 above)
Comment (malikchaya2) [reply model/task/task.go:2745]: It looks like it is only used by the resolver, so I believe yes.

File: model/task/task.go (hunk 6 above)
Comment (malikchaya2) [reply model/task/task.go:2767]: Thanks for catching this! I probably need $nor instead.

File: rest/data/task_test.go (hunk 7 above)
Comment (malikchaya2) [reply rest/data/task_test.go:85]: added all except for issues is nil. I don't know if that would ever be the case.

Comment (malikchaya2) [reply rest/data/task_test.go:85]: Issues is an array, not a pointer.

Comment (malikchaya2) [reply rest/data/task_test.go:85]: I'm adding a test for when it's empty. Please hold of reviewing.

Comment (malikchaya2) [reply rest/data/task_test.go:85]: done

File: model/task/task.go [hunk 8]
```diff
@@ -55,8 +56,9 @@ var (
 )
 
 type Task struct {
-	Id     string `bson:"_id" json:"id"`
-	Secret string `bson:"secret" json:"secret"`
+	Id     string                       `bson:"_id" json:"id"`
+	Secret string                       `bson:"secret" json:"secret"`
+	Ann    []annotations.TaskAnnotation `bson:"annotation_doc" json:"annotation_doc"`
```
Comment (ablack12) [inline model/task/task.go:61]: A couple of things here:
1) It's best to name the variable something similar to what it actually is. Ann is fine since you know what you're looking at but it isn't really a consistent variable name, better to just say "Annotations".
2) This is a list of annotation documents, so "annotation_docs" or "annotations" is more correct.
3) I think this is only used for the resolver pipeline right? So can we move this to the bottom and add a comment indicating what it's for?

**Review by ablack12 — CHANGES_REQUESTED** [review]

File: model/task/task.go (hunk 8 above)
Comment (malikchaya2) [reply model/task/task.go:61]: Oops, I put that there for testing, I meant to take that out!

Comment (ablack12) [reply model/task/task.go:61]: Actually is this a holdover from testing? Do you need the annotation document in the task struct since you do the known issue casing within the same aggregation?

**Review by ablack12 — CHANGES_REQUESTED** [review]

File: model/task/task.go (hunk 8 above)
Comment (ablack12) [reply model/task/task.go:61]: Oops race :joy:

File: model/task/task.go [hunk 9]
```diff
@@ -2741,6 +2743,35 @@ func GetTasksByVersion(versionID string, sortBy []TasksSortOrder, statuses []str
...
+											"$ne": []interface{}{
+												bson.M{
+													"$size": bson.M{"$ifNull": []interface{}{"$" + annotations.IssuesKey, []bson.M{}}},
+												}, 0,
+											},
+										},
+									},
+								},
+							}}},
+					"as": "annotation_doc",
```
Comment (ablack12) [inline model/task/task.go:2772]: nit: can we call this something plural? Doesn't really matter but will make it more clear what's happening here

File: rest/data/task_test.go [hunk 10]
```diff
@@ -81,6 +82,67 @@ func (s *TaskConnectorFetchByIdSuite) TestFindByIdAndExecution() {
...
+		Status:    evergreen.TaskFailed,
+	}
+	s.NoError(task_known.Insert())
+	s.NoError(task_not_known.Insert())
+	s.NoError(task_no_annotation.Insert())
+	s.NoError(task_with_empty_issues.Insert())
+
+	issue := annotations.IssueLink{URL: "https://issuelink.com", IssueKey: "EVG-1234", Source: &annotations.Source{Author: "chaya.malik"}}
+
+	a_with_issue := annotations.TaskAnnotation{TaskId: "task_known", TaskExecution: 0, Issues: []annotations.IssueLink{issue}}
```
Comment (ablack12) [inline rest/data/task_test.go:117]: Can we also add a test with multiple annotations for different executions to verify that we're querying execution with taskId correctly?

**Review by ablack12 — CHANGES_REQUESTED** [review]
LGTM with nit! Great work!

**Review by ablack12 — APPROVED** [review]
Meant to hit approve haha with those comments addressed

File: model/task/task.go (hunk 8 above)
Comment (malikchaya2) [reply model/task/task.go]: done. Thanks for spotting it!

File: rest/data/task_test.go (hunk 10 above)
Comment (malikchaya2) [reply rest/data/task_test.go]: done

File: model/task/task.go (hunk 9 above)
Comment (malikchaya2) [reply model/task/task.go]: done
